src/compiler/checker.ts

```diff
@@ -6703,8 +6790,11 @@ module ts {
                 }
                 else {
                     var exprType = checkExpression(<Expression>node.body);
-                    if (node.type) {
-                        checkTypeAssignableTo(exprType, getTypeFromTypeNode(node.type), node.body, /*headMessage*/ undefined);
+                    if (returnType) {
+                        if (isAsync) {
+                            exprType = getAwaitedType(exprType, /*fallbackType*/ exprType);
+                    }
+                        checkTypeAssignableTo(exprType, returnType, node.body, /*headMessage*/ undefined);

```

Comment:
To clarify, here are a few definitions:
* **promised type** - The *promised type* of a `Promise` is the type of its fulfilled value. e.g. `number` in `Promise<number>`
* **awaited type** - The *awaited type* of an expression is its *promised type* if the expression is a `Promise`; otherwise, it is the type of the expression. 

From within an async function you can return either a non-promise value or a promise. Any Promise/A+ compatible implementation will always assimilate any foreign promise, so we should not be checking assignability of a promise to the return type. Instead, we need to check assignability of the *awaited type* of the concise body against the *promised type* of its return type annotation.

```TypeScript
var fn1 = async (): WinJS.Promise<number> => 1;
var fn2 = async (): WinJS.Promise<number> => WinJS.Promise.wrap(1);
var fn3 = async (): WinJS.Promise<number> => Q.Promise.resolve(1);
```

In all three of the above examples, we need to verify that the *awaited type* of the concise body is assignable to the *promised type* of `WinJS.Promise<number>`. In all three cases, the *awaited type* of the body is `number` and the *promised type* of the return type is `number`, so the assignability relation is valid.

The `getAwaitedType` function can be used to retrieve both the *awaited type* of an expression as well as the *promised type* of a Promise. The `checkAwaitableReturnType` function is used to verify that a type is a valid Promise constructor and returns its *awaited type*.

I will add the above documentation to the `getAwaitedType` and `checkAwaitableReturnType` functions for future clarification.

The check is outside of the `if (isAsync)` block as we still need to check the assignability relation if the function is not async.